### PR TITLE
types: return an error if voting power overflows

### LINKED ISSUE #4755: Integer Overflow in ValidatorSet.VerifyCommitTrusting
There is a possible integer overflow when calculating `needed` inside `VerifyCommitTrusting`:  https://github.com/tendermint/tendermint/blob/master/types/validator_set.go#L769 

Assume that the verifying client provides a `trustValue` of `17/51` and a commit that has zero valid validator signatures. Since `TotalVotingPower` and `Numerator` are Go `int64` types, the computation `(vals.TotalVotingPower() * trustLevel.Numerator)` will overflow, returning the value −8646911284551352337. The needed value therefore evaluates to −169547280089242202. Zero validator signatures successfully verified, therefore `talliedVotingPower` is 0. 0 > `votingPowerNeeded`, thus the commit will successfully verify despite not meeting the trust requirements and containing zero validator signatures.

Recommendation is to use safe arithmetic to compute `votingPowerNeeded` (similar to `safeAdd` and `safeSub`) and to return an error if an overflow occurs.

_via the light client audit_

### PR DESCRIPTION
**Pull request by melekes**
in VerifyCommitTrusting

Closes #4755

### CODE REVIEW

File: types/validator_set.go [hunk 1]
```diff
@@ -749,11 +749,15 @@ func (vals *ValidatorSet) VerifyFutureCommit(newSet *ValidatorSet, chainID strin
 
 // VerifyCommitTrusting verifies that trustLevel ([1/3, 1]) of the validator
 // set signed this commit.
+//
 // NOTE the given validators do not necessarily correspond to the validator set
 // for this commit, but there may be some intersection.
+//
+// Panics if trustLevel is invalid.
```
Comment (tessr) [inline types/validator_set.go:756]: does this actually panic? it looks like it returns an error to me

Comment (melekes) [reply types/validator_set.go:756]: It does. Line 763 - `panic(fmt.Sprintf("trustLevel must be within [1/3, 1], given %v", trustLevel))`

File: types/validator_set.go [hunk 2]
```diff
@@ -953,3 +963,33 @@ func safeSubClip(a, b int64) int64 {
...
+
+	for absOfB > 1 {
+		c, overflow = safeAdd(c, a)
+		if overflow {
+			return c, true
+		}
+		absOfB--
+	}
+
+	if (b < 0 && a > 0) || (b < 0 && a < 0) {
```
Comment (ebuchman) [inline types/validator_set.go:990]: Can't this just be `if b < 0` ?

File: types/validator_set.go [hunk 3]
```diff
@@ -953,3 +963,33 @@ func safeSubClip(a, b int64) int64 {
 	}
 	return c
 }
+
+func safeMul(a, b int64) (int64, bool) {
```
Comment (ebuchman) [inline types/validator_set.go:967]: Would it not be more efficient to just check that `|a| < MaxInt/ |b|` like for safeAdd?

File: types/validator_set.go (hunk 2 above)
Comment (melekes) [reply types/validator_set.go:990]: no. `b < 0 && a != 0` should work though

**Pull request merged**